src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -949,6 +924,43 @@ fn preprocessed_markdown_links(s: &str) -> Vec<PreprocessedMarkdownLink> {
 }
 
 impl LinkCollector<'_, '_> {
+    fn resolve_links(&mut self, item: &Item) {
+        if !self.cx.render_options.document_private
+            && let Some(def_id) = item.item_id.as_def_id()
+            && let Some(def_id) = def_id.as_local()
+            && !self.cx.tcx.effective_visibilities(()).is_exported(def_id)
```

Comment:
Are you sure `is_exported` is the correct thing here, and not `is_reachable`? Can you add a test case for
```rust
mod private {
  /// [core::str::FromStr]
  pub struct Bar;
}

pub fn foo() -> private::Bar {
    private::Bar
}
```
and make sure it behaves the same both before and after this change?

Reply:
In both cases `Bar` is not documented and `Bar` in `pub fn foo() -> Bar` in the generated doc is not a link.

Rustdoc uses `is_reachable` in only one place when it generates json output, and I'm not sure why.